ODIN_II/SRC/generate_blif_stats.cpp

```diff
@@ -0,0 +1,1128 @@
+ * input is evaluated first.
+ *------------------------------------------------------------------*/
+void set_path_lengths(nnode_t *node, short marker)
+{
+	num_traversals_stat++;
+	nnode_t *next_node = NULL;
+	nnet_t *next_net = NULL;
+	
+	if(node->traverse_visited == marker)
+	{
+		// The idea here is to fill in gaps when a node has been left unset but visited because of a cycle, it looks ahead one to see if its neighbours have been since set
+		if(node->stat->min_dist_to_out+1 == 0)
+		{
+			single_node_lookahead(node, node->stat->traversal_id+1, marker);
+		}
+		return;
+	}
+	if(!node->stat)
+	{
+		node->stat = (nnode_stats_t *) vtr::calloc(1, sizeof(nnode_stats_t));
+		node->stat->traversal_id = 0;
+	}
+	node->stat->max_dist_to_out = 0;
+	node->stat->min_dist_to_out = 0-1;
+	unsigned int sum_dist = 0;
+	node->stat->avg_max_dist = 0;
+	unsigned int total_node_fanout = 0;
+	node->traverse_visited = marker;
+	if(node->type == OUTPUT_NODE)
+	{
+		node->stat->min_dist_to_out = node->stat->max_dist_to_out = 0;
+		return;
+	}
+	node->stat->traversal_id++;
+	for (int i = 0; i < node->num_output_pins; i++)
+	{
+		if (node->output_pins[i]->net == NULL)
+			continue;
+		next_net = node->output_pins[i]->net;
+		next_net->traversal_id++;
+		total_node_fanout += next_net->num_fanout_pins;
+		for (int j = 0; j < next_net->num_fanout_pins; j++)
+		{
+			if(next_net->fanout_pins[j] == NULL) // this happens with .v input sometimes
+				continue;
+			next_node = next_net->fanout_pins[j]->node;
+			if(next_node == NULL)
+				continue;
+			unsigned int test_max;
+			unsigned int test_min;
+			// if on path, skip
+			if(next_node->traverse_visited == marker && next_node->stat->traversal_id == next_net->traversal_id)
+			{
+				continue;
+			}
+			set_path_lengths(next_node, marker);
+			test_min = next_node->stat->min_dist_to_out;
+			if(test_min+1 == 0)
+				continue; // don't bother setting anything if next_node is itself unset
+			test_max = next_node->stat->max_dist_to_out+1;
+			sum_dist += test_max;
+			if(node->stat->min_dist_to_out > test_min)
+				node->stat->min_dist_to_out = test_min+1;
+			if(node->stat->max_dist_to_out < test_max)
+				node->stat->max_dist_to_out = test_max;
+		}
+		next_net->traversal_id = 0;
+	}
+	node->stat->traversal_id = 0;
+	node->stat->avg_max_dist = (sum_dist == 0) ? 0.0 : ((float)sum_dist / (float)total_node_fanout);
+	return;
+}
+
+/*-------------------------------------------------------------------
+ * (function: pathlen_heuristic)
+ * The idea of this is to do a dfs starting at random fanout node and run many times to see if this gives us any kind of indication of what mean path lengths should be.
+ * Ideally would probably iterate over the precreding fanout nodes after as well. Over enough iterations the results are consistent.
+ * Very very costly to just repeat this many times.	
+ *------------------------------------------------------------------*/
+int pathlen_heuristic(nnode_t *node, short marker)
```

Comment:
I would simply call it "aproximate_longest_path"  and adjust the comment to reflect that it returns an approximated max